### PR TITLE
Support bigint and large numbers especially for evm parameters and return values

### LINKED ISSUE #600: Consider more restrictive schemas for EVM base types
I was thinking about this some more. Perhaps we need to separate this schema into two parts.

1. What is allowed for blockchain RPC / smart contract calls
2. What is generally allowed

The original goal of the `paraSchema` on L9 was 2), but we are now limiting it to 1).

Solidity doesn't have floating point numbers, so limiting it to `z.number().int()` is fine. However, a `JsonApiCondition` can require floating point numbers. For example, the `paramOrContextParamSchema` is used for the value used for the returnValueTest () - and for a JSON API, checking a floating point number is perfectly fine, but now it won't be; typified by the fact that I saw you had to modify the floating point test.

_Originally posted by @derekpierre in https://github.com/nucypher/taco-web/pull/598#discussion_r1816589503_

### LINKED ISSUE #634: Support BigInt especially for EVM parameters and return values
A suggestion to add support for BigInt which need to be serialized and de-serialized at taco-web and the nodes. This is especially that int `number` cannot hold large int numbers that evm supports and return as `bigint`. 

Not handling bigints would cause currently an exception, if there was a bigint.

(The description above has been updated to point to int numbers instead of mentioning just numbers)

_Reviewers:_
@derekpierre @KPrasch

### PR DESCRIPTION
**Pull request by Muhammad-Altabba**
**Type of PR:**

- [x] Bugfix
- [ ] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:**

- [ ] 1
- [ ] 2
- [x] 3

**What this does:**

This PR is to solve [1 bug](https://github.com/nucypher/taco-web/issues/634#issuecomment-2734700957) and add [1 new feature](https://github.com/nucypher/taco-web/issues/634#issue-2912153613) by converting `bigint` and large numbers to hex strings (or smaller than the min negative number) when they are serialized. An integration test has been added which fails till TACo nodes would support this.

For TACo nodes to support this, the needed changes are made at https://github.com/nucypher/nucypher/pull/3585.

**Issues fixed/closed:**

 - Fixes #634
 - Fixes #600  (improves naming)

**Why it's needed:**

> Explain how this PR fits in the greater context of the NuCypher Network. E.g.,
> if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**

> What should reviewers focus on? Is there a particular commit/function/section
> of your PR that requires more attention from reviewers?

### CODE REVIEW

File: packages/shared/src/utils.ts [hunk 1]
```diff
@@ -44,15 +44,22 @@ const sortedReplacer = (_key: string, value: unknown) => {
...
   if (value instanceof Uint8Array) {
     return `0x${toHexString(value)}`;
   }
+  if (
+    typeof value === 'bigint' ||
+    (typeof value === 'number' &&
+      (value > Number.MAX_SAFE_INTEGER || value < Number.MIN_SAFE_INTEGER))
+  ) {
+    return `0x${value.toString(16)}`;
+  }
```
Comment (derekpierre) [inline packages/shared/src/utils.ts:57]: Is the idea here that you would convert large ints into hex?

Until now, hex has only been used for transporting bytes in JSON.  I'm concerned about nodes being able to determine the difference between hex that represents bytes vs hex that represent integers just by looking at the hex...?

File: packages/taco/test/conditions/predefined/erc721.test.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ describe('ERC721Balance', () => {
       chain: TEST_CHAIN_ID,
       returnValueTest: {
         comparator: '==',
-        value: '10',
+        value: BigInt(10),
```
Comment (derekpierre) [inline packages/taco/test/conditions/predefined/erc721.test.ts:47]: This test only tests that BigInts were not previously accepted. It's not really testing that a large value would have failed in the past.

File: packages/shared/src/utils.ts (hunk 1 above)
Comment (Muhammad-Altabba) [reply packages/shared/src/utils.ts:57]: Yes, I propose to convert to hex string. Except if there is some issue at an edge case. https://github.com/nucypher/taco-web/issues/634#issuecomment-2737114631

File: packages/taco/test/conditions/predefined/erc721.test.ts (hunk 2 above)
Comment (Muhammad-Altabba) [reply packages/taco/test/conditions/predefined/erc721.test.ts:47]: Yes, exactly. This test is for the new feature added (which is to support bigint).

File: packages/shared/src/utils.ts (hunk 1 above)
Comment (cygnusv) [reply packages/shared/src/utils.ts:57]: > Is the idea here that you would convert large ints into hex?
> 
> Until now, hex has only been used for transporting bytes in JSON.  I'm concerned about nodes being able to determine the difference between hex that represents bytes vs hex that represent integers just by looking at the hex...?

Well that's not for the nodes to discern, but it's dictated by the ABI (or reference schema in the case of RPC methods).

File: packages/taco/test/conditions/lingo.test.ts [hunk 3]
```diff
@@ -165,18 +164,21 @@ describe('check large numbers', () => {
...
+
     const contractCondition = new ContractCondition(
       uint256ContractConditionProps,
     );
     const conditionExpr = new ConditionExpression(contractCondition);
     expect(conditionExpr.toJson()).toContain(
-      '115792089237316195423570985008687907853269984665640564039457584007913129639935',
+      (115792089237316195423570985008687907853269984665640564039457584007913129639935).toString(
+        16,
+      ),
```
Comment (Muhammad-Altabba) [inline packages/taco/test/conditions/lingo.test.ts:176]: We would use '1157.....9935' if we adopted converting the large numbers to strings. But I think converting them to hex string is better because it start with `0x`.

File: packages/shared/src/utils.ts (hunk 1 above)
Comment (derekpierre) [reply packages/shared/src/utils.ts:57]: Need to think on this some more.

For `returnValueTest` we use `ast.literal_eval(...)` on both result and execution value, which will convert values into common types. This applies to other conditions as well. We added code to handle bytes as hex strings because bytes aren't allowed in JSON - https://github.com/nucypher/nucypher/pull/3303. So an int represented by a hex string is probable fine.

However for `parameters` (or `params` for non-blockchain conditions) we pass values as provided, instead of trying to perform conversion of the values. 

In general, we've typically avoided type conversions, because we aren't always going to have an ABI to check (non-onchain conditions). For something like a `JsonRPCCondition` that doesn't have an ABI, the node can't do conversions on its behalf and it expects parameter values to be passed in the types they are supposed to be in. If you pass in a string for an int when an int is expected, that may/may not fail when sending that param to the endpoint. 

It also prevents us from having inconsistent condition definitions (various values for the same parameter) that still work, and potentially complex code for type checking and type conversions. The creator of the condition understands best the correct types, so we expect them to provide the correct types. 

In web3py's case it does some internal type conversion on our behalf and the check is https://web3py.readthedocs.io/en/stable/web3.main.html#web3.w3.is_encodable.

For example, to pass bytes to web3py you can pass it as python `bytes` or as hex string:
```python
In [69]: w3.is_encodable("bytes", b"somebytes")
Out[69]: True

In [70]: w3.is_encodable("bytes", "0x736f6d656279746573")
Out[70]: True
```

However, if the value is not encodable as the expected type then it fails. So for uint256, while you can pass an int, you can't pass a hex string:
```python
In [72]: w3.is_encodable("uint256", "0x1234")
Out[72]: False

In [73]: w3.is_encodable("uint256", 115792089237316195423570985008687907853269984665640564039457584007913129639935)
Out[73]: True
```

So if a hex string is provided to the node for a contract parameter that expects uint256, the condition execution would fail due to a type mismatch.

Comment (derekpierre) [reply packages/shared/src/utils.ts:57]: Potentially having the int as a string could be simpler than hex. These `param` / `parameter` / `value`are open fields without type restrictions. Then at least I can check if the string can be converted to an int, and if not leave it as a string.

File: packages/taco/test/conditions/predefined/erc721.test.ts (hunk 2 above)
Comment (derekpierre) [reply packages/taco/test/conditions/predefined/erc721.test.ts:47]: I really meant that based on the commit message of - https://github.com/nucypher/taco-web/pull/637/commits/d8773d2d152942f045f8043ef7d92e2d0cae97da i.e. "modify a test that passes now but would fail before the PR" -> this would have failed because of a type mismatch.

I think that instead, you can add an additional test to confirm the ability to use BigInt instead of modifying an existing one - keeping original tests are good for ensuring regression testing. For new functionality we should add new tests.

File: packages/shared/src/utils.ts (hunk 1 above)
Comment (derekpierre) [reply packages/shared/src/utils.ts:57]: Here's a possibility. For any numbers outside of the safe range, we use strings in the JSON, and numbers within the safe range (bigint or regular int) we continue to use them as numbers in the JSON - https://github.com/nucypher/taco-web/compare/epic-v0.6.x...derekpierre:taco-web:large-number-tests. 

Everything works as before for numbers provided within the safe range. For bigints (outside of the safe number range) they will be strings, and on the node side, we'll check if the string is convertible to an int, and if so convert it, if not, it'll just remain a string. Need to investigate the node changes required.

Thoughts?

Comment (Muhammad-Altabba) [reply packages/shared/src/utils.ts]: Thanks @derekpierre for the helpful discussion.

And regarding:
> Everything works as before for numbers or bigints provided within the safe range. For bigints (outside of the safe number range) they will be strings, and on the node side, we'll check if the string is convertible to an int, and if so convert it, if not, it'll just remain a string. Need to investigate the node changes required.


This is also what this PR is about except some bit minor details:
- I think hex string are bit better as checking the at the nodes would be: if the string start with `0x` try parse into a number (wether this check is with if or at a regex). Also hex string are shorter than number string. Also in all cases wether the large number is a string or a hex string, a type conversion is needed for uint256.

> w3.is_encodable("uint256", "115792089237316195423570985008687907853269984665640564039457584007913129639935")
Gives: False.
> w3.is_encodable("uint256", "0x11111")
Gives: False.
> w3.is_encodable("uint256", 115792089237316195423570985008687907853269984665640564039457584007913129639935)
Gives: True.

- Using Zod transform is nice. But I lean toward the separation between data validation and data serialization. By using Zod to specify the schema accepted form the user. And keep the serialization conversion to the time before sending the data. I think there is no right and wrong here. But I lean toward decoupling definition from transformation.

- Using number for the safe range of bigint sounds also good. I updated the code accordingly.

Comment (derekpierre) [reply packages/shared/src/utils.ts]: > I think hex string are bit better as checking the at the nodes would be: if the string start with 0x try parse into a number (wether this check is with if or at a regex

Bytes are also added to the JSON as hex strings as well (that's the original `u8ToHexReplacer` code), so for any hex string (starts with `0x`) it can either be a hex representation of bytes or now a hex representation of an int, but the node can't tell the difference because parsing a hex string as either an int or bytes will work, and the resulting type matters.

```python
In [4]: int("0xdeadbeef", 16)
Out[4]: 3735928559

In [5]: bytes(HexBytes("0xdeadbeef"))
Out[5]: b'\xde\xad\xbe\xef'
```

As opposed to having it be string in the JSON where the node can do the following:
```python
try:
    result = int(value)
    return result
except ValueError:
    # regular string
    pass
    
return value
```
i.e. either the parsing of the string to an int works because it was a number, or it's just a regular string.


> Using Zod transform is nice. But I lean toward the separation between data validation and data serialization. By using Zod to specify the schema accepted form the user. And keep the serialization conversion to the time before sending the data. I think there is no right and wrong here. But I lean toward decoupling definition from transformation.

That's fair.

File: packages/taco/src/conditions/schemas/context.ts [hunk 4]
```diff
@@ -11,12 +11,17 @@ export const contextParamSchema = z
...
+const paramSchema = z.union([
+  plainStringSchema,
+  z.boolean(),
+  z.number(),
+  z.bigint(),
+]);
 
 const nonFloatParamSchema = z
-  .union([plainStringSchema, z.boolean(), z.number().int()])
+  .union([plainStringSchema, z.boolean(), z.number().int(), z.bigint()])
```
Comment (derekpierre) [inline packages/taco/src/conditions/schemas/context.ts:22]: I think we need to refine this bigint, because it shouldn't allow values > uint256 max or values < int256 min for onchain conditions - see https://github.com/derekpierre/taco-web/blob/large-number-tests/packages/taco/src/conditions/schemas/context.ts#L35. 

That may beg the follow-up question as to whether `nonFloatParamSchema` is still the best name for this schema.

File: packages/taco/test/conditions/context.test.ts [hunk 5]
```diff
@@ -122,7 +123,9 @@ describe('context', () => {
       const condition = new RpcCondition(testRpcConditionObj);
       const conditionContext = new ConditionContext(condition);
       expect(
-        JSON.stringify(condition.toObj()).includes(USER_ADDRESS_PARAM_DEFAULT),
+        JSON.stringify(condition.toObj(), toHexReplacer).includes(
```
Comment (derekpierre) [inline packages/taco/test/conditions/context.test.ts:126]: I think for here and in other places, anywhere that we are calling `JSON.stringify(...)` directly, should be replaced by the call to `toJSON(...)` from utils.

File: packages/taco/src/conditions/schemas/context.ts (hunk 4 above)
Comment (Muhammad-Altabba) [reply packages/taco/src/conditions/schemas/context.ts]: Good point!
Also checked for the int number to be in the same range.

File: packages/taco/integration-test/encrypt-decrypt.test.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    BigInt(10000000000000000), // 0.01 ETH/PLY
+  ])(
+    'should encrypt and decrypt a message with condition value %s',
+    async (value) => {
+      // Create test message
+      const messageString = 'This is a secret 🤐';
+      const message = toBytes(messageString);
+
+      // Create condition
+      const hasPositiveBalance = new conditions.base.rpc.RpcCondition({
```
Comment (derekpierre) [inline packages/taco/integration-test/encrypt-decrypt.test.ts:84]: I think you can just do a single `CompoundCondition` (`AND`) comprised of multiple rpc conditions each using the different values. Then you can just run this once.

Also I would say once of those should be a check for a balance < uint256 max i.e.
```typescript
returnValueTest: {
    comparator: "<",
    value: BigInt("115792089237316195423570985008687907853269984665640564039457584007913129639935"),
}
```

Comment (Muhammad-Altabba) [reply packages/taco/integration-test/encrypt-decrypt.test.ts:84]: The suggestion on using `CompoundCondition` I think it is great for the case.
And yes, it would be overkill to repeat the full test just for different condition values. I basically made it like this to make it easy for this to be checked with the nodes as part of the issue scope definition.

Probably for an integration test it is good to use extensive complicated conditions that covers almost all conditions. Would be good to craft such when doing https://github.com/nucypher/taco-web/issues/638.

File: packages/taco/test/conditions/lingo.test.ts [hunk 7]
```diff
@@ -132,3 +134,94 @@ describe('check that valid lingo in python is valid in typescript', () => {
...
 });
+
+describe('check large numbers serialization into hex string', () => {
+  const numbersToTest = [
+    {
+      name: 'uint256 large positive',
+      // The max uint256 cannot be encoded correctly in javascript becuase it is too large.
+      // For that some arbitraty extra large positive number is used:
+      //  11579208923731619542357098500868790785326998466564056403945758400791312963993
+      // Which is less by a littile from the max uint256 (note 5 at the end):
```
Comment (manumonti) [inline packages/taco/test/conditions/lingo.test.ts]:
```suggestion
      // Which is less by a little from the max uint256 (note 5 at the end):
```

**Review by manumonti — APPROVED** [review]
LGTM! I have just catched a small typo, but not a blocker

**Comment by derekpierre** [comment]
> Thanks @derekpierre for your nice contributions.
> 
> I have no more thoughts to apply to this PR. Feel free to approve and merge or let me know if you would like me to add something.

👍 . Thank you as well - it was a nice tag-team effort.

Nothing to add, just looking for reviews from others. My review is implicit since I added commits.
